pkg/daytona/unmarshal.go

```diff
@@ -0,0 +1,282 @@
+package daytona
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/hashicorp/vault/api"
+)
+
+const (
+	tagVaultPathKeyName     = "vault_path_key"
+	tagVaultPathDataKeyName = "vault_path_data_key"
+
+	tagVaultDataKeyName = "vault_data_key"
+
+	defaultDataKeyFieldName = "value"
+)
+
+var (
+	// ErrValueInput indicates the provided value is not a struct pointer
+	ErrValueInput = errors.New("the provided value must be a struct pointer")
+)
+
+// SecretUnmarshler reads data from Vault and stores the result(s) in the
+// a provided struct. This can be useful to inject sensitive configuration
+// items directly into config structs
+type SecretUnmarshler struct {
+	client      *api.Client
+	tokenString string
+	tokenFile   string
+}
+
+// NewSecretUnmarshler returns a new SecretUnmarshler, applying any options
+// that are supplied.
+func NewSecretUnmarshler(opts ...Option) (*SecretUnmarshler, error) {
+	var s SecretUnmarshler
+	for _, opt := range opts {
+		opt.Apply(&s)
+	}
+
+	if s.client == nil {
+		client, err := api.NewClient(api.DefaultConfig())
+		if err != nil {
+			return nil, fmt.Errorf("failed to create new vault client: %w", err)
+		}
+		s.client = client
+	}
+
+	if s.tokenString != "" && s.tokenFile != "" {
```

Comment:
what happens if neither are passed in?  

Also is it possible to have this use the ENV vars normally set for Daytona to auto auth to Vault and get a token?

Reply:
`api.DefaultConfig()`, which is passed to the default client, attempts to extract address, token and other vault-specific configurations from environment variables. If someone passes in their own client, I'm presuming they configured it before hand.